src/util/invariant.h

```diff
@@ -51,14 +53,63 @@ Author: Martin Brain, martin.brain@diffblue.com
 ** CPROVER_INVARIANT_* macros.
 */
 
+/// A logic error, augmented with a distinguished field to hold a backtrace.
 class invariant_failedt: public std::logic_error
 {
+  /// Stores the backtrace set by `invariant_violated`.
+  std::string backtrace;
+
 public:
-  explicit invariant_failedt(const std::string& what) : logic_error(what) {}
-  explicit invariant_failedt(const char* what) : logic_error(what) {}
+  invariant_failedt(
+    const std::string &what):
+    logic_error(what),
+    backtrace()
+  {
+  }
+
+  /// Set the backtrace stored against this error report.
+  /// This is overwritten by `invariant_violated`, which is called
+  /// from the `INVARIANT` family of macros.
+  void set_backtrace(const std::string &_backtrace)
```

Comment:
I would consider having this private and making `invariant_violated` a friend or a static member of this class.